specs/safe-liveness-checking.md

```diff
@@ -0,0 +1,119 @@
+signer. This time is made publicly available by calling a `lastLive(address)(Timestamp)` method.
+
+Signers may also call the contract's `showLiveness()()` method directly in order to prove liveness.
+
+### The Liveness Module
+
+A `LivenessModule` is also created which does the following:
+
+1. Has a function `removeOwner()` that anyone may call to specify an owner to be removed from the
+   Safe.
+1. The Module would then check the `LivenessGuard.lastLive()` to determine if the signer is
+   eligible for removal.
+1. If so, it will call the Safe's `removeSigner()` to remove the non-live signer, and if necessary
+   reduce the threshold.
+1. When a member is removed, the signing parameters are modified such that `M/N` is the lowest ratio
+   which remains greater than or equal to 75%. Using integer math, this can be expressed as `M = (N * 75 + 99) / 100`.
+
+### Owner removal call flow
+
+The following diagram illustrates the flow for removing a single owner. The `verifyFinalState`
+box indicates calls to the Safe which ensure the final state is valid.
+
+```mermaid
+sequenceDiagram
+    participant User
+    participant LivenessModule
+    participant LivenessGuard
+    participant Safe
+    User->>LivenessModule: removeOwners([previousOwner], [owner])
+    LivenessModule->>LivenessGuard: lastLive(owner)
+    LivenessModule->>Safe: getOwners()
+    LivenessModule->>Safe: removeOwner(previousOwner, owner)
+
+    alt verifyFinalState
+    LivenessModule->>Safe: getOwners()
+    LivenessModule->>Safe: getThreshold()
+    LivenessModule->>Safe: getGuard()
+    end
+```
+
+### Shutdown
+
+In the unlikely event that the signer set (`N`) is reduced below the allowed threshold, then (and only then) is a
+   shutdown mechanism activated which removes the existing signers, and hands control of the
+   multisig over to a predetermined entity.
+
+### Security Properties
+
+The following security properties must be upheld:
+
+1. Signatures are assigned to the correct signer.
+1. Non-signers are unable to create a record of having signed.
+1. A signer cannot be censored or grieffed such that their signing is not recorded.
+1. Signers may demonstrate liveness either by signing a transaction or by calling directly to the
+   guard.
+1. The module only removes a signer if they have demonstrated liveness during the interval, or
+     if necessary to convert the safe to a 1 of 1.
+1. The module sets the correct 75% threshold upon removing a signer.
+1. During a shutdown the module correctly removes all signers, and converts the safe to a 1 of 1.
+
+### Interdependency between the Guard and Module
+
+The Guard has no dependency on the Module, and can be used independently to track liveness of
+Safe owners. The Module however does have a dependency on the Guard, only one guard contract can
+be set on the Safe, and the Module will be unable to function if the Guard is removed.
+
+### Deployment
```

Comment:
Thanks, this is helpful. Can we add similar steps for removing and/or updating the guard and/or module?

Reply:
20aaed5dbed362445a825f9828f714e43230afbc